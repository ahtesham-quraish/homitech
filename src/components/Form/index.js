import React, { createContext, Component } from "react";
import validations from "../../utils/validations";
import { isEmpty } from "lodash";
import { ERROR, SUCCESS } from "../../constants";

export const FormCtx = createContext({
  fields: {},
  errors: {},
  validateClasses: {},
  isFetching: false,
});

export default class Form extends Component {
  state = {
    fields: {},
    errors: {},
    validateClasses: {},
    isFetching: false,
  };
  render() {
    const { fields, errors, validateClasses, isFetching } = this.state;
    const formCtx = {
      fields,
      errors,
      validateClasses,
      isFetching,
      addField: (data) => {
        this.addField(data);
      },
      setFields: this.setFields,
      validateField: this.validateField,
      clearValidationField: this.clearValidationField,
      validateForm: this.validateForm,
      setFormData: this.setFormData,
    };
    const { formClass } = this.props.classes;
    return (
      <form action='' className={formClass} autoComplete='off'>
        <FormCtx.Provider value={formCtx}>
          {this.props.children}
        </FormCtx.Provider>
      </form>
    );
  }
  componentDidUpdate(preProps) {
    const { apiResponse } = this.props;
    if (preProps.apiResponse !== apiResponse && apiResponse === 200) {
      this.setState(
        {
          errors: {},
          validateClasses: {},
          isFetching: false,
        },
        this.clearFields
      );
    }
  }
  clearFields = () => {
    const { fields } = this.state;
    for (const unit in fields) {
      const field = fields[unit];
      if (field) {
        if (field.type === "file") {
          continue;
        }
        this.addField({
          field: {
            ...field,
            value: "",
          },
        });
      }
    }
  };

  setFields = (event, { id }) => {
    if (event && event.persist) {
      event.persist();
    }
    const { fields } = this.state;
    const field = fields[id];

    this.addField({
      field: {
        ...field,
        value: event && event.currentTarget ? event.currentTarget.value : event,
      },
    });
  };

  addField = ({ field }) => {
    const { id } = field;

    field = {
      value: "",
      ...field,
    };

    if (id) {
      this.setState((prevState) => {
        const defulatClasses =
          prevState.fields &&
          prevState.fields[id] &&
          prevState.fields[id].classes
            ? prevState.fields[id].classes
            : {};
        const classes = { ...defulatClasses, ...field.custom };
        if (!isEmpty(defulatClasses)) {
          field.classes = classes;
        }
        return {
          ...prevState,
          fields: {
            ...prevState.fields,
            [id]: field,
          },
        };
      });

      return;
    }

    throw new Error(`please add 'id' field to the input: ${field}`);
  };
  setFormData = (data) => {
    this.setState((prevState) => ({
      ...prevState,
      ...data,
    }));
  };
  clearValidationField = (id) => {
    const {
      value: fieldValue,
      validate,
      displayName,
      customRules = {},
    } = this.state.fields[id];
    let clearError = true;
    const rules = validate ? validate.split("|") : "";
    if (rules.length) {
      for (const rule in rules) {
        const ruleName = rules[rule];
        const ruleValue = this.state.fields[id][ruleName];
        const validation = validations[ruleName] || customRules[ruleName];
        const isRuleSatisfied =
          ruleName !== "required" && !fieldValue
            ? true
            : validation.rule(ruleValue).test(fieldValue.toString());
        if (!isRuleSatisfied) {
          clearError = false;
        }
      }
      this.setState((prevState) => ({
        ...prevState,
        errors: {
          ...prevState.errors,
          [id]: clearError ? "" : prevState.errors[id],
        },
        validateClasses: {
          ...prevState.validateClasses,
          [id]: clearError ? SUCCESS : prevState.validateClasses[id],
        },
      }));
    }
  };
  validateForm = (fieldName) => {
    const { shouldValidateForm, fields } = this.state;
    const field = fields[fieldName];
    if (field) {
      if (field.shouldValidateField) {
        this.validateField(field);
      }

      return;
    }

    for (const unit in fields) {
      const fieldData = fields[unit];

      if (fieldData && fieldData.validation) {
        this.validateField(unit);
      }
    }

    return;
  };
  validateField = (id) => {
    let error = "";
    const {
      value: fieldValue,
      validate,
      displayName,
      customRules = {},
    } = this.state.fields[id];
    const rules = validate ? validate.split("|") : "";

    if (rules.length) {
      for (const rule in rules) {
        const ruleName = rules[rule];
        const ruleValue = this.state.fields[id][ruleName];
        const validation = validations[ruleName] || customRules[ruleName];
        const isRuleSatisfied =
          ruleName !== "required" && !fieldValue
            ? true
            : validation.rule(ruleValue).test(fieldValue.toString());
        if (!isRuleSatisfied) {
          error = validation.formatter.apply(null, [displayName || id]);
        }
        if (error !== "") {
          break;
        }
      }
      this.setState((prevState) => ({
        ...prevState,
        errors: {
          ...prevState.errors,
          [id]: error,
        },
        validateClasses: {
          ...prevState.validateClasses,
          [id]: error !== "" ? ERROR : SUCCESS,
        },
      }));
    }
  };
}
Form.defaultProps = {
  events: {},
  classes: {
    formClass: "form",
  },
};
