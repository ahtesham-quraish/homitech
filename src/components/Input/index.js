import React, { useContext, useEffect } from "react";
import { FormCtx } from "../Form";
import { ERROR, SUCCESS } from "../../constants";
// Generic component for input 
const Input = (props) => {
  const { id } = props;
  const {
    setFields,
    addField,
    fields,
    validateField,
    clearValidationField,
    errors,
    validateClasses,
  } = useContext(FormCtx);
  const field = fields[id] || {};
  const {
    name,
    rows,
    value,
    validate,
    isOnBlur,
    validation,
    placeholder,
    type = "text",
    events = {},
    classes = {},
  } = field;
  const fieldError = errors[id];
  const { onChange, onBlur, ...restEvents } = events;
  const {
    inputClass,
    fieldClass,
    errorClass,
    defaultClass,
    customClasses,
    customContainerClasses,
  } = classes;
  const handleChange = (event) => {
    try {
      setFields(event, field);
    } catch (error) {
      throw error;
    }

    if (typeof onChange === "function") {
      onChange({
        ...field,
        value: event.target.value,
      });
    }
  };
  const handleOnBlur = (event) => {
    try {
      setFields(event, field);
    } catch (error) {
      throw error;
    }
    if (field.isOnBlur && validation) {
      validateField(field.id);
    }
  };

  useEffect(() => {
    if (value !== undefined && !isOnBlur && validation) {
      validateField(id);
    }
    if (value !== undefined && validation) {
      clearValidationField(id);
    }
  }, [value, id]);

  useEffect(() => {
    addField({
      field: props,
      value,
    });
  }, []);

  const fieldProps = {
    ...restEvents,
    id,
    name,
    type,
    value,
    validate,
    placeholder,
    className: fieldClass,
    onChange: handleChange,
    onBlur: handleOnBlur,
  };

  if (type === "textarea") {
    delete fieldProps.type;
    delete fieldProps.value;

    fieldProps.defaultValue = value;
    fieldProps.rows = rows || 2;
  }
  return field && field.value !== undefined ? (
    <div className={`${defaultClass} ${customContainerClasses}`}>
      {type === "textarea" ? (
        <textarea {...fieldProps} />
      ) : (
        <input
          key={field.componentKey}
          {...fieldProps}
          className={`${inputClass} ${
            field && field.validation ? validateClasses[field.id] : ""
          }  ${customClasses}`}
        />
      )}
      <div className={errorClass}>{fieldError}</div>
    </div>
  ) : (
    ""
  );
};
Input.defaultProps = {
  events: {},
  classes: {
    defaultClass: "form-group",
    contClass: "",
    errorClass: "form-error-label",
    inputClass: "form-control",
    labelClass: "form-label",
    customContainerClasses: "",
    customClasses: "",
  },
};
export default Input;
