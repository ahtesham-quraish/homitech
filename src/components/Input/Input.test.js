import React, { createContext } from "react";
import renderer, { update, create, act } from "react-test-renderer";
import PropTypes from "prop-types"; // eslint-disable-line import/no-extraneous-dependencies
import { mount } from "enzyme";
import Input from "./index";
import Form from "../Form";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

/* eslint-disable react/jsx-filename-extension */
const defaultContext = {
  addField: jest.fn(),
  setFieldValue: jest.fn(),
  validateForm: jest.fn(),
  formData: {
    defaultClasses: {
      contClass: "default-contClass",
      fieldClass: "default-fieldClass",
      errorClass: "default-errorClass",
      labelClass: "default-labelClass",
    },
    fields: {
      email: {
        value: "",
      },
    },
  },
};
const contextTypes = {
  addField: PropTypes.func,
  setFieldValue: PropTypes.func,
  formData: PropTypes.object,
  validateForm: PropTypes.func,
};
const defaultInputArgs = {
  id: "email",
  type: "text",
  value: "123",
  onFieldChange: null,
  validate: "required",
  placeholder: "enter value",
  displayName: "Email",
  shouldUseDefaultClasses: true,
  classes: {
    contClass: "cont-class",
    fieldClass: "field-class",
    errorClass: "error-class",
    labelClass: "label-class",
  },
};

// Snapshot matching for Input Coponent
describe(">>> Input Container -- Snapshot Test", () => {
  it("Matches the snapshot with default props", () => {
    let tree;

    act(() => {
      tree = create(
        <Form>
          <Input id='email' />
        </Form>
      );
    });
    act(() => {
      tree.update(
        <Form>
          <Input id='email' />
        </Form>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it("Matches the snapshot with predefined value", () => {
    let tree;
    act(() => {
      tree = create(
        <Form>
          <Input id='email' />
        </Form>
      );
    });
    act(() => {
      tree.update(
        <Form>
          <Input id='email' value='test' />
        </Form>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("Matches the snapshot with custom props", () => {
    const customProps = { ...defaultInputArgs };

    customProps.placeholder = "custom placeholder";
    customProps.displayName = "custom display name";
    let tree;
    act(() => {
      tree = create(
        <Form defaultClasses={customProps.defaultClasses}>
          <Input {...customProps} />
        </Form>
      );
    });
    act(() => {
      tree.update(
        <Form defaultClasses={customProps.defaultClasses}>
          <Input {...customProps} />
        </Form>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
// *************************************************************

/* eslint-enable */

// Check DOM Behaviour

describe(">>> Input Container -- Shallow Rendering", () => {
  let mountedScreen;
  const inputScreen = (customProps = {}, customContext = {}) => {
    const initialState = {
      id: "email",
      value: "",
    };

    if (!mountedScreen) {
      const updatedState = {
        ...initialState,
        ...customProps,
      };

      mountedScreen = mount(
        <Form>
          <Input {...updatedState} />
        </Form>
      );
    }

    return mountedScreen;
  };

  beforeEach(() => {
    mountedScreen = undefined;
  });

  it("Triggers input change function when input changed", () => {
    const props = {
      events: {
        onChange: jest.fn(),
      },
    };
    const tree = inputScreen(props);

    const input = tree.find("input").at(0);

    input.simulate("change", {
      target: {
        value: "test",
      },
    });

    expect(props.events.onChange).toHaveBeenCalled();
  });
});
//
