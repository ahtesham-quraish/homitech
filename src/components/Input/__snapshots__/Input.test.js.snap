// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Input Container -- Snapshot Test Matches the snapshot with custom props 1`] = `
<form
  action=""
  autoComplete="off"
  className="form"
>
  <div
    className="undefined undefined"
  >
    <input
      className="undefined   undefined"
      id="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="custom placeholder"
      type="text"
      validate="required"
      value="123"
    />
    <div
      className="error-class"
    />
  </div>
</form>
`;

exports[`>>> Input Container -- Snapshot Test Matches the snapshot with default props 1`] = `
<form
  action=""
  autoComplete="off"
  className="form"
>
  <div
    className="form-group "
  >
    <input
      className="form-control   "
      id="email"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="form-error-label"
    />
  </div>
</form>
`;

exports[`>>> Input Container -- Snapshot Test Matches the snapshot with predefined value 1`] = `
<form
  action=""
  autoComplete="off"
  className="form"
>
  <div
    className="form-group "
  >
    <input
      className="form-control   "
      id="email"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="form-error-label"
    />
  </div>
</form>
`;
